core/src/block_creation_loop.rs

```diff
@@ -0,0 +1,521 @@
+    // AND that replay no longer touches poh recorder. At this point BlockCreationLoop is the sole
+    // modifier of poh_recorder
+    {
+        let mut w_poh_recorder = ctx.poh_recorder.write().unwrap();
+        w_poh_recorder.enable_alpenglow();
+    }
+    reset_poh_recorder(&ctx.bank_forks.read().unwrap().working_bank(), &ctx);
+
+    while !ctx.exit.load(Ordering::Relaxed) {
+        // Check if set-identity was called at each leader window start
+        if my_pubkey != cluster_info.id() {
+            let my_old_pubkey = my_pubkey;
+            my_pubkey = cluster_info.id();
+            ctx.my_pubkey = my_pubkey;
+
+            warn!(
+                "Identity changed from {my_old_pubkey} to {my_pubkey} during block creation loop"
+            );
+        }
+
+        // Wait for the voting loop to notify us
+        let LeaderWindowInfo {
+            start_slot,
+            end_slot,
+            parent_block: (parent_slot, _),
+            skip_timer,
+        } = {
+            let window_info = leader_window_notifier.window_info.lock().unwrap();
+            let (mut guard, timeout_res) = leader_window_notifier
+                .window_notification
+                .wait_timeout_while(window_info, Duration::from_secs(1), |wi| wi.is_none())
+                .unwrap();
+            if timeout_res.timed_out() {
+                continue;
+            }
+            guard.take().unwrap()
+        };
+
+        trace!("Received window notification for {start_slot} to {end_slot} parent: {parent_slot}");
+        if let Err(e) = produce_window(start_slot, end_slot, parent_slot, skip_timer, &mut ctx) {
+            // Give up on this leader window
+            error!(
+                "{my_pubkey}: Unable to produce window {start_slot}-{end_slot}, skipping window: \
+                 {e:?}"
+            );
+            continue;
+        }
+    }
+
+    info!("{my_pubkey}: Block creation loop shutting down");
+}
+
+/// Resets poh recorder
+fn reset_poh_recorder(bank: &Arc<Bank>, ctx: &LeaderContext) {
+    trace!("{}: resetting poh to {}", ctx.my_pubkey, bank.slot());
+    assert!(ctx.record_receiver.is_shutdown() && ctx.record_receiver.is_safe_to_restart());
+    let next_leader_slot = ctx.leader_schedule_cache.next_leader_slot(
+        &ctx.my_pubkey,
+        bank.slot(),
+        bank,
+        Some(ctx.blockstore.as_ref()),
+        GRACE_TICKS_FACTOR * MAX_GRACE_SLOTS,
+    );
+
+    ctx.poh_recorder
+        .write()
+        .unwrap()
+        .reset(bank.clone(), next_leader_slot);
+}
+
+/// Produces the leader window from `start_slot` -> `end_slot` using parent
+/// `parent_slot` while abiding to the `skip_timer`
+fn produce_window(
+    start_slot: Slot,
+    end_slot: Slot,
+    parent_slot: Slot,
+    skip_timer: Instant,
+    ctx: &mut LeaderContext,
+) -> Result<(), StartLeaderError> {
+    // Insert the first bank
+    start_leader_retry_replay(start_slot, parent_slot, skip_timer, ctx)?;
```

Comment:
could we move this to the top of the while loop instead? Then we could:

1. get rid of the second check at the bottom of the loop
2. make the while loop conditional on `!ctx.exit.load(Ordering::Relaxed) && slot > end_slot`